### PR TITLE
feat: overlay images.

### PR DESCRIPTION
**Pull request by biggest-dave**
Run locally and hit with: 
http://localhost:8080/__origami/service/image/v2/images/raw/https%3A%2F%2Fd1e00ek4ebabms.cloudfront.net%2Fproduction%2Fa88c2b59-ac2d-4f70-a68a-57b8499bd70f.jpg?source=next-article&fit=scale-down&quality=highest&width=700&dpr=1&overlay=https://ft-next-assets-prod.s3.eu-west-1.amazonaws.com/assets/ft-banner.png&overlay_height=0.5&overlay_y=30

You should see: 

Screenshot 2023-10-26 at 10 01 37

### CODE REVIEW

File: lib/transformers/cloudinary.js [hunk 1]
```diff
@@ -50,11 +50,42 @@ function buildCloudinaryTransforms(transform) {
...
-	return cloudinaryTransforms;
+
+	const overlay = transform.getOverlay();
+
+
+	if (overlay) {
+
+		const overlayPosition = transform.getOverlayPosition();
+		const overlayHeight = transform.getOverlayWidth();
+		const overlayWidth = transform.getOverlayHeight();
```
Comment (notlee) [inline lib/transformers/cloudinary.js:66]: Oops! Even when I found the problem, it took me so long to see that this was the cause 🙈 

```suggestion
		const overlayHeight = transform.getOverlayHeight();
		const overlayWidth = transform.getOverlayWidth();
```

File: lib/image-transform.js [hunk 2]
```diff
@@ -68,6 +87,147 @@ module.exports = class ImageTransform {
...
+	}
+
+	/**
+	 * Set the overlay width
+	 * @param {String} value 
+	 */
+
+	setOverlayWidth(value) {
+		// the cloudinary api will ignore an empty string
+		this.overlay_width = value ? value : '';
```
Comment (notlee) [inline lib/image-transform.js:138]: Reviewing your PR is getting me more (re?) familiarised with the Image Service which is good 😄 Looks like `lib/image-transform.js` is intended as an abstraction layer. Should Cloudinary specific fallbacks be moved to `lib/transformers/cloudinary.js`?

File: lib/image-transform.js [hunk 3]
```diff
@@ -68,6 +87,147 @@ module.exports = class ImageTransform {
...
+		 * Get the overlay x.
+		 * @return {String}
+		 */
+	getOverlayX() {
+		return this.overlay_x;
+	}
+
+	/**
+	 * Set the overlay x
+	 * @param {String} value
```
Comment (notlee) [inline lib/image-transform.js:170]: Could/should this be `{string|number}`?

File: lib/image-transform.js [hunk 4]
```diff
@@ -49,6 +57,17 @@ module.exports = class ImageTransform {
...
+		
+		if (properties.overlay) {
+			this.setOverlay(properties.overlay);
+			this.setOverlayPosition(properties.overlay_position);
+			this.setOverlayWidth(properties.overlay_width);
+			this.setOverlayHeight(properties.overlay_height);
+			this.setOverlayX(properties.overlay_x);
+			this.setOverlayY(properties.overlay_y);
+			this.setOverlayGravity(properties.overlay_gravity);
+			this.setOverlayCrop(properties.overlay_crop);
```
Comment (notlee) [inline lib/image-transform.js:69]: For each of these, could we limit options here to a known set? Maybe even just the options you need right now? See `setGravity`. That way we have a smaller, more well defined task if we opt to move to a new supplier next year (but it's also quite easy to allow extra Cloudinary features).

File: lib/image-transform.js [hunk 5]
```diff
@@ -68,6 +87,147 @@ module.exports = class ImageTransform {
 		this.uri = ImageTransform.sanitizeUriValue(value, errorMessage);
 	}
 
+		/**
+	 * Get the overlay.
+	 * @return {String}
+	 */
+		getOverlay() {
+			return this.overlay;
```
Comment (notlee) [inline lib/image-transform.js:95]: Little indenting issue here (urgh, we need to set up a code formatter!)

**Review by notlee — CHANGES_REQUESTED** [review]
Looking good! One little bug and couple of suggestions here with our future maintenance in mind 🙏

**Comment by notlee** [comment]
Also, bonus points! Apologies if I didn't think to mention before. Cache can be tricky with the image service. I suspect it would be a welcome ability to be able to "purge all images with x as their overlay" when we update an overlay's source image. To do that we'll want to set a [Surrogate Key](https://developer.fastly.com/reference/http/http-headers/Surrogate-Key/) on responses for Fastly.

I think that would happen as [part of the proxy response handler](https://github.com/Financial-Times/origami-image-service/blob/b1655858b90bf1455564f734431686030e5ef5e0/lib/image-service.js#L179). We'd want to differentiate an image used as an overlay vs. the same image being requested directly, so we'd maybe want to add something like `encodeKeys('overlay-' + overlayUrl)` conditionally, when an overlay is requested.

File: lib/transformers/cloudinary.js (hunk 1 above)
Comment (biggest-dave) [reply lib/transformers/cloudinary.js:66]: Well spotted!

File: lib/image-transform.js (hunk 4 above)
Comment (biggest-dave) [reply lib/image-transform.js]: I took out position and crop. I left gravity in as I could image that being used very soon (we're having debates about if it should come in from the south_east or south_west).

Comment (notlee) [reply lib/image-transform.js]: Then we can document the more limited set of values explicitly, rather than need people to reference Cloudinary docs.

Comment (biggest-dave) [reply lib/image-transform.js]: Done!

File: lib/image-transform.js [hunk 6]
```diff
@@ -654,6 +794,37 @@ module.exports = class ImageTransform {
...
+			'fit',
+			'imagga_crop',
+			'imagga_scale',
+			'lfill',
+			'limit',
+			'lpad',
+			'mfit',
+			'mpad',
+			'pad',
+			'thumb'
```
Comment (notlee) [inline lib/image-transform.js:824]: Can we reduce this list to just what you need for now? It'll make our potential transition to a new supplier easier

Comment (biggest-dave) [reply lib/image-transform.js]: done

**Review by notlee — APPROVED** [review]
Looks good! 🥳 🎉

**Comment by origamiserviceuser** [comment]
:tada: This PR is included in version v453.10.0 :tada:

**Pull request merged**